private/library.rkt

```diff
@@ -0,0 +1,296 @@
+
+(define (removable-directory? lib dir)
+  (and (hash-has-key? (library-lib (library-data-table lib)) dir)
+       (not (equal? user-script-dir dir))))
+
+(define (add-directory lib dir)
+  (define data (library-lib lib))
+  (struct-copy library lib
+               [lib (struct-copy library-data data
+                                 [table (hash-update (library-data-table data)
+                                                     dir
+                                                     values
+                                                     setalw)])]))
+
+(define (remove-directory lib dir)
+  (define data (library-lib lib))
+  (struct-copy library lib
+               [lib (struct-copy library-data data
+                                 [table (hash-remove (library-data-table data)
+                                                     dir)])]))
+
+(define (in/exclude set-change lib dir filename)
+  (define data (library-lib lib))
+  (struct-copy library lib
+               [lib (if (set-member? (library-collects-dirs lib) dir)
+                        (struct-copy
+                         library-data data
+                         [collection-based-exclusions
+                          (set-add (library-data-collection-based-exclusions data)
+                                   (path->normalized-lib-module-path lib (build-path dir filename)))])
+                        (struct-copy
+                         library-data data
+                         [table (hash-update (library-data-table data)
+                                             filename
+                                             (λ (excludes)
+                                               (set-add excludes filename)))]))]))
+
+(define (exclude lib dir filename)
+  (in/exclude set-add lib dir filename))
+
+(define (include lib dir filename)
+  (in/exclude set-remove lib dir filename))
+
+(provide/contract
+ [library?            (any/c . -> . boolean?)]
+ #;
+ [new-library         (-> library?)]
+ [load                (-> library?)]
+ [save!               (library? . -> . void?)]
+ [directories         ([library?]
+                       [#:sorted? any/c]
+                       . ->* . (listof path?))]
+ [directory->enabled+file (library?
+                           path?
+                           . -> . (listof (cons/c boolean? path-element?)))]
+ [all-enabled-scripts (library? . -> . (listof path?))]
+ [directory<?         (library? path? path? . -> . boolean?)]
+ [directory->pretty-string ([library? path?]
+                            [#:file (or/c #f path-element?)]
+                            . ->* . immutable-string?)]
+ [path->writable-module-path (library?
+                              path?
+                              . -> . (and/c (list/c (or/c 'file 'lib) immutable-string?)
+                                            module-path?))]
+ [add-directory       (library?
+                       (and/c path? absolute-path?)
```

Comment:
I wonder if these contracts should use `complete-path?` instead of `absolute-path?`. (This is not a change: the single-commit diff is somewhat clearer about that.)

Reply:
Oh, I didn't know that absolute paths can omit the drive letter on Windows! o_O

Reply:
I've changed these to use `complete-path?`.